### PR TITLE
#2685 - Import Provincial Restrictions from SFAS to SIMS (Part 2) - PR 1

### PR DESCRIPTION
**Pull request by sh16011993**
### As a part of this PR, the following have been completed:

- [x] At student account creation, for the restrictions related to the newly created student, `processed` in `sfas_restrictions` has is set to false. This will ensure that the next run of the `sfas-integration` scheduler inserts the appropriate restrictions related to this student in the `students_restrictions` table.
- [x] Content for the activity page is updated as per the following: **Title: Code LGCY - Legacy Restriction;** **Content: You have a historical restriction on your account. Please contact StudentAid BC for additional assistance.**
- [x] When a LGCY restriction is automatically applied into SIMS via the SFAs bridge, automatic email is created to the Ministry staff (studentaidbc.gov.bc.ca) notifying them of such a restriction - include the same basic identifying details as all other emails. **Subject line: New Legacy Restriction - [last name];**

**Note: e2e tests to be part of the next PR.**

### Screenshots:

**Account Activity Page:** 
image

**GC Notify Template:**
image

**Email Notification:** 
image

### CODE REVIEW

File: sources/packages/backend/libs/integrations/src/services/sfas/sfas-restriction.service.ts [hunk 1]
```diff
@@ -108,6 +124,58 @@ export class SFASRestrictionService
     await this.repo.save(restriction, { reload: false, transaction: false });
   }
 
+  /**
+   * Updates processed to false for the SFAS restrictions corresponding to the provided SFAS Individual Id.
+   * @param individualId provided SFAS Individual Id.
+   */
+  async updateProcessedStatus(individualId: number): Promise<void> {
+    await this.repo.update({ individualId }, { processed: false });
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/services/sfas/sfas-restriction.service.ts:132]: I do not see this simple update as essential to be part of this service.
Keeping it in the student service only and changing it as mentioned in the comment below will prevent many files from being changed along this PR because the dependency will no longer be needed for the student service.
https://github.com/bcgov/SIMS/pull/2910#discussion_r1509423750

File: sources/packages/backend/libs/integrations/src/services/sfas/sfas-restriction.service.ts [hunk 2]
```diff
@@ -108,6 +124,58 @@ export class SFASRestrictionService
...
+    const studentLegacyRestrictions = await this.studentRestrictionRepo.find({
+      select: {
+        student: {
+          birthDate: true,
+          user: { id: true, firstName: true, lastName: true, email: true },
+        },
+      },
+      relations: { student: { user: true } },
+      where: {
+        restriction: { restrictionCode: "LGCY" },
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/services/sfas/sfas-restriction.service.ts:154]: I believe that a student ID would be enough to execute the query.

File: sources/packages/backend/libs/integrations/src/services/sfas/sfas-restriction.service.ts [hunk 3]
```diff
@@ -108,6 +124,58 @@ export class SFASRestrictionService
...
+  }
+
+  /**
+   * Create notifications for newly created legacy student restrictions for the ministry.
+   * @param studentIds student ids for which the legacy restriction notification needs to be sent.
+   * @param entityManager entity manager to execute in transaction.
+   */
+  private async createLegacyRestrictionNotifications(
+    studentIds: number[],
+    entityManager?: EntityManager,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/services/sfas/sfas-restriction.service.ts:142]: If an entity manager is a provider it should be used.
If there is no reason for this `entityManager` to be optional it can be changed and used to query the student data.

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts [hunk 4]
```diff
@@ -261,6 +262,43 @@ export class NotificationActionsService {
...
+      messageType: NotificationMessageType.LegacyRestrictionAdded,
+      messagePayload: {
+        email_address: this.gcNotifyService.ministryToAddress(),
+        template_id: templateId,
+        personalisation: {
+          givenNames: notification.firstName ?? "",
+          lastName: notification.lastName,
+          studentEmail: notification.email,
+          dob: getDateOnlyFormat(notification.birthDate),
+          "date/time": this.getDateTimeOnPSTTimeZone(),
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:290]: Is there a reason to create a variable name as `date/time` instead of `"dateTime"` following the same for `givenNames`, `studentEmail`.

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts [hunk 5]
```diff
@@ -261,6 +262,43 @@ export class NotificationActionsService {
...
+    );
+    const notificationsToSend = notifications.map((notification) => ({
+      userId: notification.userId,
+      messageType: NotificationMessageType.LegacyRestrictionAdded,
+      messagePayload: {
+        email_address: this.gcNotifyService.ministryToAddress(),
+        template_id: templateId,
+        personalisation: {
+          givenNames: notification.firstName ?? "",
+          lastName: notification.lastName,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:287]: This template is replicating an issue from other templates where the given names are completely concatenated with the last name without leaving a space between then, as shown in the PR description. Can we add the space to the template?

File: sources/packages/backend/libs/src-sql/sfas-restrictions/Bulk-insert-legacy-restrictions.sql [hunk 6]
```diff
@@ -54,4 +54,4 @@ WHERE
       student_restrictions.student_id = filtered_students.student_id
       AND student_restrictions.restriction_id = $1
       AND student_restrictions.is_active = TRUE
-  );
\ No newline at end of file
+  ) RETURNING student_id;
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/src-sql/sfas-restrictions/Bulk-insert-legacy-restrictions.sql:57]: ❤️

File: sources/packages/forms/src/form-definitions/studentaccountactivity.json [hunk 7]
```diff
@@ -10394,8 +10435,5 @@
       "id": "e6z1qd9",
       "addons": []
     }
-  ],
-  "created": "2023-01-11T21:27:07.204Z",
-  "modified": "2023-04-21T19:39:43.414Z",
-  "machineName": "studentAccountActivity"
+  ]
```
Comment (andrewsignori-aot) [inline sources/packages/forms/src/form-definitions/studentaccountactivity.json:10438]: Thanks for the cleanup 👍

**Review by andrewsignori-aot — COMMENTED** [review]
Great work, please take a look at the comments.

File: sources/packages/backend/apps/api/src/services/student/student.service.ts [hunk 8]
```diff
@@ -216,6 +219,15 @@ export class StudentService extends RecordDataModelService<Student> {
         auditUserId,
         entityManager,
       );
+      // For the newly created student, update the Processed status in the
+      // SFAS Restrictions table to false. This will enable for the SFAS
+      // Restrictions from previous imports that were originally marked as
+      // processed to be re-processed when the SFAS Integration Scheduler
+      // runs again, thus causing the restrictions imported from SFAS to be
+      // applied to the newly created student account.
+      await this.sfasRestrictionService.updateProcessedStatus(
```
[the remark this answers is not in the document]
Comment (guru-aot) [reply sources/packages/backend/apps/api/src/services/student/student.service.ts:228]: @andrewsignori-aot  you mean having the status update when the student creation is happening around line 199? If yes i do agree.

File: sources/packages/backend/libs/integrations/src/services/sfas/sfas-restriction.service.ts [hunk 9]
```diff
@@ -71,6 +79,14 @@ export class SFASRestrictionService
             { processed: false },
             { processed: true, updatedAt: new Date() },
           );
+        const bulkInsertLegacyRestrictionStudentIds =
```
Comment (guru-aot) [inline sources/packages/backend/libs/integrations/src/services/sfas/sfas-restriction.service.ts:82]: This is happening in a transaction as intended 👍

File: sources/packages/backend/libs/integrations/src/services/sfas/sfas-restriction.service.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/integrations/src/services/sfas/sfas-restriction.service.ts]: The second line of the comment above was suggesting to use the student directly instead of the `StudentRestriction`. The query can be based on `getRepository(StudentRestriction)`, unless I am missing something.

File: sources/packages/backend/libs/integrations/src/services/sfas/sfas-restriction.service.ts [hunk 10]
```diff
@@ -108,6 +122,50 @@ export class SFASRestrictionService
...
+        },
+        relations: { student: { user: true } },
+        where: {
+          student: { id: In(studentIds) },
+        },
+      });
+    if (!studentLegacyRestrictions?.length) {
+      // There are no notifications to be sent.
+      return;
+    }
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/services/sfas/sfas-restriction.service.ts:152]: If the below comment is accepted I do not see a reason to keep this check,
https://github.com/bcgov/SIMS/pull/2910#discussion_r1511943923

File: sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts [hunk 11]
```diff
@@ -261,6 +262,43 @@ export class NotificationActionsService {
...
+    );
+    const notificationsToSend = notifications.map((notification) => ({
+      userId: notification.userId,
+      messageType: NotificationMessageType.LegacyRestrictionAdded,
+      messagePayload: {
+        email_address: this.gcNotifyService.ministryToAddress(),
+        template_id: templateId,
+        personalisation: {
+          givenNames: notification.firstName ?? "",
+          lastName: (notification.firstName ? " " : "") + notification.lastName,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/services/src/notifications/notification/notification-actions.service.ts:287]: Please revert this change. The request was to fix the template.
The current code will give the `lastName` a useless extra white space.

**Review by andrewsignori-aot — COMMENTED** [review]
Thank you for making the extra changes, please take a look at the comments.

**Review by andrewsignori-aot — APPROVED** [review]
Thank you for making the changes, looks good 👍

File: sources/packages/backend/apps/api/src/services/student/student.service.ts [hunk 12]
```diff
@@ -216,6 +218,15 @@ export class StudentService extends RecordDataModelService<Student> {
...
       );
+      // For the newly created student, update the Processed status in the
+      // SFAS Restrictions table to false. This will enable for the SFAS
+      // Restrictions from previous imports that were originally marked as
+      // processed to be re-processed when the SFAS Integration Scheduler
+      // runs again, thus causing the restrictions imported from SFAS to be
+      // applied to the newly created student account.
+      await entityManager
+        .getRepository(SFASRestriction)
+        .update({ individualId: sfasIndividual.id }, { processed: false });
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/api/src/services/student/student.service.ts:229]: As discussed, please add the if condition with sfas individual here.

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @sh16011993

**Review by dheepak-aot — APPROVED** [review]
Looks good 👍

I reviewed here as I noticed the if condition. Still the secondary reviewer is not done for the PR.

**Review by lewischen-aot — APPROVED** [review]
LGTM! Great work @sh16011993
